File: combinatorics/resource/TMD/graph_tree/Code/problem04/problem04.py
from typing import List, Tuple, Dict, Optional

# =========================================================================
# ==                      DATA STRUCTURE DEFINITIONS                     ==
# =========================================================================

# Intermediate representation: A list of tuples [(u, v, weight), ...]
EdgeList = List[Tuple[int, int, int]]

# 1. Adjacency Matrix
AdjacencyMatrix = List[List[Optional[int]]]

# 2. Adjacency List (for multigraphs/general graphs)
AdjacencyList = Dict[int, List[Tuple[int, int]]]

# 3. Adjacency Map (for simple graphs)
AdjacencyMap = Dict[int, Dict[int, int]]

# --- Tree Representations ---
# 1. Array of Parents
ArrayOfParents = List[int]

# 2. First-Child, Next-Sibling
FCNS_Representation = List[Dict[str, int]]


# =========================================================================
# ==                   GRAPH & TREE CONVERTER FUNCTIONS                  ==
# =========================================================================

# --- Graph Converters ---

def edge_list_to_adj_matrix(edge_list: EdgeList, num_nodes: int) -> AdjacencyMatrix:
    matrix = [[None] * num_nodes for _ in range(num_nodes)]
    for u, v, weight in edge_list:
        if u < num_nodes and v < num_nodes:
            matrix[u][v] = weight
    return matrix

def adj_matrix_to_edge_list(matrix: AdjacencyMatrix) -> EdgeList:
    edge_list = []
    num_nodes = len(matrix)
    for i in range(num_nodes):
        for j in range(num_nodes):
            if matrix[i][j] is not None:
                edge_list.append((i, j, matrix[i][j]))
    return edge_list

def edge_list_to_adj_list(edge_list: EdgeList, num_nodes: int) -> AdjacencyList:
    adj_list = {i: [] for i in range(num_nodes)}
    for u, v, weight in edge_list:
        if u in adj_list:
            adj_list[u].append((v, weight))
    return adj_list

def adj_list_to_edge_list(adj_list: AdjacencyList) -> EdgeList:
    edge_list = []
    for u, neighbors in adj_list.items():
        for v, weight in neighbors:
            edge_list.append((u, v, weight))
    return edge_list

def edge_list_to_adj_map(edge_list: EdgeList) -> AdjacencyMap:
    adj_map = {}
    for u, v, weight in edge_list:
        if u not in adj_map:
            adj_map[u] = {}
        adj_map[u][v] = weight
    return adj_map
    
def adj_map_to_edge_list(adj_map: AdjacencyMap) -> EdgeList:
    edge_list = []
    for u, neighbors in adj_map.items():
        for v, weight in neighbors.items():
            edge_list.append((u, v, weight))
    return edge_list

# --- Tree Converters ---

def array_of_parents_to_tree_edges(parents: ArrayOfParents) -> Tuple[EdgeList, int]:
    edges = []
    root = -1
    for i, p in enumerate(parents):
        if p != -1:
            edges.append((p, i, 1))  # Assuming unweighted tree
        else:
            root = i
    return edges, root

def tree_edges_to_array_of_parents(edges: EdgeList, num_nodes: int) -> ArrayOfParents:
    parents = [-1] * num_nodes
    for p, c, _ in edges:
        parents[c] = p
    return parents

def tree_edges_to_fcns(edges: EdgeList, num_nodes: int) -> FCNS_Representation:
    children_list = [[] for _ in range(num_nodes)]
    for p, c, _ in edges:
        children_list[p].append(c)

    fcns = [{'first_child': -1, 'next_sibling': -1} for _ in range(num_nodes)]
    for i in range(num_nodes):
        if children_list[i]:
            fcns[i]['first_child'] = children_list[i][0]
            for j in range(len(children_list[i]) - 1):
                current_child = children_list[i][j]
                next_child = children_list[i][j + 1]
                fcns[current_child]['next_sibling'] = next_child
    return fcns

def fcns_to_tree_edges(fcns: FCNS_Representation) -> Tuple[EdgeList, int]:
    edges = []
    num_nodes = len(fcns)
    is_child = [False] * num_nodes
    for parent_idx, node_data in enumerate(fcns):
        child = node_data['first_child']
        while child != -1:
            edges.append((parent_idx, child, 1))
            if 0 <= child < num_nodes:
                is_child[child] = True
            child = fcns[child]['next_sibling']
    
    root = -1
    for i in range(num_nodes):
        if not is_child[i]:
            root = i
            break
            
    return edges, root

# =========================================================================
# ==                           HELPER PRINTERS                           ==
# =========================================================================

def print_edge_list(name: str, edges: EdgeList):
    print(f"{name}: {sorted(edges)}")

def print_matrix(name: str, matrix: AdjacencyMatrix):
    print(f"{name}:")
    for row in matrix:
        print(f"  {[str(v) if v is not None else 'N' for v in row]}")

def print_adj_list(name: str, adj_list: AdjacencyList):
    print(f"{name}:")
    for u, neighbors in sorted(adj_list.items()):
        print(f"  {u}: {neighbors}")
        
def print_adj_map(name: str, adj_map: AdjacencyMap):
    print(f"{name}:")
    for u, neighbors in sorted(adj_map.items()):
        print(f"  {u}: {neighbors}")

# =========================================================================
# ==                            MAIN FUNCTION                            ==
# =========================================================================

if __name__ == "__main__":
    # --- GRAPH DEMO ---
    num_graph_nodes = 4
    graph_edges = [(0, 1, 5), (0, 2, 3), (1, 2, 2), (1, 3, 6), (2, 3, 1)]

    print("=" * 60)
    print("       GRAPH CONVERTER DEMONSTRATION (Problem 4)        ")
    print("=" * 60)
    
    print_edge_list("Original Edge List", graph_edges)
    
    print("\n--- Testing Matrix Conversions ---")
    matrix = edge_list_to_adj_matrix(graph_edges, num_graph_nodes)
    print_matrix("To Adjacency Matrix", matrix)
    edges_from_matrix = adj_matrix_to_edge_list(matrix)
    print_edge_list("Back to Edge List", edges_from_matrix)

    print("\n--- Testing Adjacency List Conversions ---")
    adj_list = edge_list_to_adj_list(graph_edges, num_graph_nodes)
    print_adj_list("To Adjacency List", adj_list)
    edges_from_adj_list = adj_list_to_edge_list(adj_list)
    print_edge_list("Back to Edge List", edges_from_adj_list)
    
    print("\n--- Testing Adjacency Map Conversions ---")
    adj_map = edge_list_to_adj_map(graph_edges)
    print_adj_map("To Adjacency Map", adj_map)
    edges_from_adj_map = adj_map_to_edge_list(adj_map)
    print_edge_list("Back to Edge List", edges_from_adj_map)

    # --- TREE DEMO ---
    num_tree_nodes = 7
    # 0 is root, 1 and 2 are its children, 3 and 4 are children of 1, etc.
    parents_array = [-1, 0, 0, 1, 1, 2, 2]

    print("\n" + "=" * 60)
    print("        TREE CONVERTER DEMONSTRATION (Problem 4)        ")
    print("=" * 60)
    
    print(f"Original Array of Parents: {parents_array}")

    print("\n--- Testing Array of Parents <-> Edge List ---")
    tree_edges, root = array_of_parents_to_tree_edges(parents_array)
    print_edge_list("To Edge List", tree_edges)
    print(f"Detected Root: {root}")
    parents_from_edges = tree_edges_to_array_of_parents(tree_edges, num_tree_nodes)
    print(f"Back to Array of Parents: {parents_from_edges}")

    print("\n--- Testing Edge List -> FCNS ---")
    fcns_repr = tree_edges_to_fcns(tree_edges, num_tree_nodes)
    print("To First-Child Next-Sibling Representation:")
    for i, node in enumerate(fcns_repr):
        print(f"  Node {i}: {node}")

    print("\n--- Testing FCNS -> Edge List ---")
    edges_from_fcns, root_from_fcns = fcns_to_tree_edges(fcns_repr)
    print_edge_list("Back to Edge List", edges_from_fcns)
    print(f"Detected Root: {root_from_fcns}")
    
    print("\n" + "=" * 60)

File: combinatorics/resource/TMD/graph_tree/Code/problem04/problem04.cpp
#include <iostream>
#include <vector>
#include <map>
#include <optional>
#include <utility> // For std::pair
#include <algorithm> // For std::sort (for comparison)
#include <numeric>


// Run with C++ 23
// =========================================================================
// ==                      DATA STRUCTURE DEFINITIONS                     ==
// =========================================================================

// Intermediate representation: A simple struct for an edge
struct Edge {
    int u;
    int v;
    int weight = 1;

    // For sorting and comparison
    bool operator<(const Edge& other) const {
        if (u != other.u) return u < other.u;
        if (v != other.v) return v < other.v;
        return weight < other.weight;
    }
};

// 1. Adjacency Matrix
using AdjacencyMatrix = std::vector<std::vector<std::optional<int>>>;

// 2. Adjacency List
using AdjacencyList = std::vector<std::vector<std::pair<int, int>>>;

// 3. Adjacency Map
using AdjacencyMap = std::map<int, std::map<int, int>>;

// --- Tree Representations ---
// 1. Array of Parents
using ArrayOfParents = std::vector<int>;

// 2. First-Child, Next-Sibling
struct FCNS_Node { int first_child = -1; int next_sibling = -1; };
using FCNS_Representation = std::vector<FCNS_Node>;

// =========================================================================
// ==                        GRAPH CONVERTER CLASS                        ==
// =========================================================================

class GraphConverter {
public:
    // --- From Edge List (to other formats) ---
    static AdjacencyMatrix toMatrix(const std::vector<Edge>& edge_list, int num_nodes) {
        AdjacencyMatrix matrix(num_nodes, std::vector<std::optional<int>>(num_nodes, std::nullopt));
        for (const auto& edge : edge_list) {
            if (edge.u < num_nodes && edge.v < num_nodes) {
                matrix[edge.u][edge.v] = edge.weight;
            }
        }
        return matrix;
    }

    static AdjacencyList toAdjList(const std::vector<Edge>& edge_list, int num_nodes) {
        AdjacencyList adj_list(num_nodes);
        for (const auto& edge : edge_list) {
            if (edge.u < num_nodes) {
                adj_list[edge.u].push_back({edge.v, edge.weight});
            }
        }
        return adj_list;
    }

    static AdjacencyMap toAdjMap(const std::vector<Edge>& edge_list) {
        AdjacencyMap adj_map;
        for (const auto& edge : edge_list) {
            adj_map[edge.u][edge.v] = edge.weight;
        }
        return adj_map;
    }

    // --- To Edge List (from other formats) ---
    static std::vector<Edge> fromMatrix(const AdjacencyMatrix& matrix) {
        std::vector<Edge> edge_list;
        for (size_t i = 0; i < matrix.size(); ++i) {
            for (size_t j = 0; j < matrix[i].size(); ++j) {
                if (matrix[i][j].has_value()) {
                    edge_list.push_back({(int)i, (int)j, matrix[i][j].value()});
                }
            }
        }
        return edge_list;
    }

    static std::vector<Edge> fromAdjList(const AdjacencyList& adj_list) {
        std::vector<Edge> edge_list;
        for (size_t u = 0; u < adj_list.size(); ++u) {
            for (const auto& pair : adj_list[u]) {
                edge_list.push_back({(int)u, pair.first, pair.second});
            }
        }
        return edge_list;
    }

    static std::vector<Edge> fromAdjMap(const AdjacencyMap& adj_map) {
        std::vector<Edge> edge_list;
        for (const auto& [u, neighbors] : adj_map) {
            for (const auto& [v, weight] : neighbors) {
                edge_list.push_back({u, v, weight});
            }
        }
        return edge_list;
    }
};

// =========================================================================
// ==                           HELPER PRINTERS                           ==
// =========================================================================

void printEdgeList(const std::vector<Edge>& edges) {
    std::cout << "{ ";
    for (const auto& e : edges) {
        std::cout << "(" << e.u << "," << e.v << "," << e.weight << ") ";
    }
    std::cout << "}\n";
}

void printMatrix(const AdjacencyMatrix& matrix) {
    for (const auto& row : matrix) {
        for (const auto& val : row) {
            if (val.has_value()) {
                std::cout << val.value() << "\t";
            } else {
                std::cout << "N\t";
            }
        }
        std::cout << "\n";
    }
}

void printAdjList(const AdjacencyList& adjList) {
    for (size_t i = 0; i < adjList.size(); ++i) {
        std::cout << i << ": ";
        for (const auto& p : adjList[i]) {
            std::cout << "->(" << p.first << "," << p.second << ") ";
        }
        std::cout << "\n";
    }
}

void printAdjMap(const AdjacencyMap& adjMap) {
    for (const auto& [u, neighbors] : adjMap) {
        std::cout << u << ": ";
        for (const auto& [v, weight] : neighbors) {
            std::cout << "->(" << v << "," << weight << ") ";
        }
        std::cout << "\n";
    }
}


// =========================================================================
// ==                            MAIN FUNCTION                            ==
// =========================================================================
int main() {
    int num_nodes = 4;
    // Our intermediate representation
    std::vector<Edge> original_edges = {{0, 1, 5}, {0, 2, 3}, {1, 2, 2}, {1, 3, 6}, {2, 3, 1}};
    
    std::cout << "=================================================" << std::endl;
    std::cout << "       GRAPH CONVERTER DEMO (Problem 4)        " << std::endl;
    std::cout << "=================================================\n" << std::endl;
    
    std::cout << "Original Edge List:" << std::endl;
    printEdgeList(original_edges);
    
    std::cout << "\n--- Edge List -> Adjacency Matrix ---" << std::endl;
    AdjacencyMatrix mat = GraphConverter::toMatrix(original_edges, num_nodes);
    printMatrix(mat);
    
    std::cout << "\n--- Adjacency Matrix -> Edge List ---" << std::endl;
    std::vector<Edge> edges_from_mat = GraphConverter::fromMatrix(mat);
    // Sort for consistent comparison
    std::sort(edges_from_mat.begin(), edges_from_mat.end());
    printEdgeList(edges_from_mat);
    
    std::cout << "\n--- Edge List -> Adjacency List ---" << std::endl;
    AdjacencyList alist = GraphConverter::toAdjList(original_edges, num_nodes);
    printAdjList(alist);
    
    std::cout << "\n--- Adjacency List -> Edge List ---" << std::endl;
    std::vector<Edge> edges_from_alist = GraphConverter::fromAdjList(alist);
    std::sort(edges_from_alist.begin(), edges_from_alist.end());
    printEdgeList(edges_from_alist);
    
    std::cout << "\n--- Edge List -> Adjacency Map ---" << std::endl;
    AdjacencyMap amap = GraphConverter::toAdjMap(original_edges);
    printAdjMap(amap);

    std::cout << "\n--- Adjacency Map -> Edge List ---" << std::endl;
    std::vector<Edge> edges_from_amap = GraphConverter::fromAdjMap(amap);
    std::sort(edges_from_amap.begin(), edges_from_amap.end());
    printEdgeList(edges_from_amap);

    std::cout << "\n=================================================" << std::endl;

    return 0;
}